gtsfm/densify/mvsnets/methods/PatchmatchNet/models/patchmatch.py

```diff
@@ -0,0 +1,891 @@
+
+                depth_sample = 1.0 / torch.cat(depth_clamped, dim=0)
+                del depth_clamped
+
+                return depth_sample
+
+
+class Propagation(nn.Module):
+    """ Propagation module implementation"""
+
+    def __init__(self, neighbors: int = 16):
+        """init methods
+        Args:
+            neighbors: number of neighbor views
+        """
+        super(Propagation, self).__init__()
+        self.neighbors = neighbors
+
+    def forward(
+        self,
+        batch: int,
+        height: int,
+        width: int,
+        depth_sample: torch.Tensor,
+        grid: torch.Tensor,
+        depth_min: int,
+        depth_max: int,
+        depth_interval_scale: float,
+    ) -> torch.Tensor:
+        # [B,D,H,W]
+        """forward function
+        Args:
+            batch: batch size,
+            height: depth map height,
+            width: depth map width,
+            depth_sample: sample depth map,
+            grid: grid for bilinear gridding,
+            depth_min: minimum virtual depth,
+            depth_max: maximum virtual depth,
+            depth_interval_scale: depth virtual interval scale,
+        """
+        num_depth = depth_sample.size()[1]
+        propogate_depth = depth_sample.new_empty(batch, num_depth + self.neighbors, height, width)
+        propogate_depth[:, 0:num_depth, :, :] = depth_sample
+
+        propogate_depth_sample = F.grid_sample(
+            depth_sample[:, num_depth // 2, :, :].unsqueeze(1), grid, mode="bilinear", padding_mode="border"
+        )
+        del grid
+        propogate_depth_sample = propogate_depth_sample.view(batch, self.neighbors, height, width)
+
+        propogate_depth[:, num_depth:, :, :] = propogate_depth_sample
+        del propogate_depth_sample
+
+        # sort
+        propogate_depth, _ = torch.sort(propogate_depth, dim=1)
+
+        return propogate_depth
+
+
+class Evaluation(nn.Module):
+    """Evaluation Class"""
+
+    def __init__(self, G: int = 8, stage: int = 3, evaluate_neighbors: int = 9, iterations: int = 2):
+        """initialize
+        Args:
+            G: number of groups
+            stage: stage id
+            evaluate neighbors: number of evaluate neighbors
+            iterations: number of evaluation iteration
+        """
+        super(Evaluation, self).__init__()
+
+        self.iterations = iterations
+
+        self.G = G
+        self.stage = stage
+        if self.stage == 3:
+            self.pixel_wise_net = PixelwiseNet(self.G)
+
+        self.similarity_net = SimilarityNet(self.G, evaluate_neighbors)
+
+    def forward(
+        self,
+        ref_feature: torch.Tensor,
+        src_features: List[torch.Tensor],
+        ref_proj: torch.Tensor,
+        src_projs: List[torch.Tensor],
+        depth_sample: torch.Tensor,
+        depth_min: int,
+        depth_max: int,
+        iter: iter,
+        grid: torch.Tensor = None,
+        weight: torch.Tensor = None,
+        view_weights: List = None,

```

Comment:
thanks for all the updates, Ren.


view_weights is List[np.ndarray] or List[torch.Tensor]?

Reply:
Oh, I missed one possible "if" branch, view_weights should be None or torch.Tensor